### ISSUE TITLE
Attempt to prune branches in Matches as an expression optimization 

### ISSUE THREAD

**Issue by snoble** [johnynek/bosatsu#606]
I the LetFree code there is an implementation of maybebind that can determine if a LetFreeExpression always, never, or sometimes matches a branch in a Match.
https://github.com/johnynek/bosatsu/blob/0338e369c2fcd3a3e62862547a8db15e1cb0527a/core/src/main/scala/org/bykn/bosatsu/LetFree.scala#L268

This could probably be used to eliminate branches, or even reduce down to a single branch, in the Expr.

**Comment by johnynek** [comment]
Yeah, totally. I would really like to do that, and more symbolic evaluation in general.

**Comment by johnynek** [comment]
Looked at Matchless.scala and TypedExprNormalization.scala. The match-branch pruning / optimization work requested here landed in:
- #1050 (TypedExprNormalization: maybeEvalMatch + match normalization pruning)
- #1519 (Matchless: pattern-matrix compiler to share/prune match tests)
- #1536 (Matchless: totality-aware matrix compilation to skip redundant checks)

These cover the optimization suggested in this issue, so closing.
